Support min_gtid annotation

kirs opened the pull request:
This PR is an attempt to get https://github.com/sysown/proxysql/issues/2444 done. I'm probably missing a lot, and this is the first time I do C++, so I'd really appreciate the feedback.

How does it work?

* This relies on ProxySQL tracking GTID of each backend, using https://github.com/sysown/proxysql_mysqlbinlog. That must be running on each MySQL host.
* Each client keeps min_gtid of the data that it has written (maybe even leveraging `session_track_gtids`): in a cookie/session for web requests, and in Redis for batch jobs.
* `min_gtid` is passed in a comment to ProxySQL, which lets it choose any read replica that satisfies that GTID.

https://github.com/sysown/proxysql/issues/2444#issuecomment-565388325 has a diagram showing how that can be leveraged by a typical web app.

Questions:
* What would be the best way to test this? I'm not sure I'm 100% following how ProxySQL test suite works.

I'm looking forward to your feedback ❤️

kirs committed:
Support min_gtid comment

alpes214 commented on lib/Query_Processor.cpp:
```diff
@@ -2779,6 +2779,15 @@ bool Query_Processor::query_parser_first_comment(Query_Processor_Output *qpo, ch
+				size_t vs = strlen(value);
+				if (vs > 0) {
+					char *buf=(char*)malloc(vs);
+					strncpy(buf, value, vs - 1);
```
The `size_t strlen(const char* s)` function calculates the length of the string pointed to by its parameter, excluding the terminating null byte ('\0'). Therefor `buf` contains truncated value.

alpes214 commented on lib/MySQL_Session.cpp:
```diff
@@ -6410,22 +6410,24 @@ void MySQL_Session::handler___client_DSS_QUERY_SENT___server_DSS_NOT_INITIALIZED
-				int l = index(_gtid_from_backend->gtid_uuid,':') - _gtid_from_backend->gtid_uuid;
-				trxid = strtoull(index(_gtid_from_backend->gtid_uuid,':')+1, NULL, 10);
+			if (gtid_uuid != NULL) {
+				int l = index(gtid_uuid,':') - gtid_uuid;
```
If the invalid value of gtid is passed into min_gtid, the proxysql crashes.
For example, the following query crashes proxysql
```
select /* min_gtid=111 */ @@hostname
```
The `char *index(const char* s, int c);` function returns a pointer to the matched character or NULL if the character is not found. The index return value should be checked for NULL.

alpes214 replied:
Event min_gtid is not correct and wrong value is used to find server, the proxysql should not crash. The client will be just timed out. And it looks like correct behavior

alpes214 commented on lib/MySQL_Session.cpp:
```diff
@@ -6410,22 +6410,24 @@ void MySQL_Session::handler___client_DSS_QUERY_SENT___server_DSS_NOT_INITIALIZED
-				trxid = strtoull(index(_gtid_from_backend->gtid_uuid,':')+1, NULL, 10);
+			if (gtid_uuid != NULL) {
+				int l = index(gtid_uuid,':') - gtid_uuid;
+				trxid = strtoull(index(gtid_uuid,':')+1, NULL, 10);
```
`index()` can return NULL here.

alpes214 reviewed (changes requested):
@kirs according to the logic the min_gtid takes precedence over gtid_from_hostgroup which is configured in query rules. It looks like correct behavior.

There several issues that should be resolved before merge.

kirs committed:
Validate gtid input

kirs commented:
Thanks for review @val214! I pushed another commit that adds GTID validation, to make sure that we only set `qpo->min_gtid` to legit value.

pondix commented:
Automated message: PR pending admin approval for build testing

alpes214 commented:
@kirs, thank you for your changes but there is an issue that still should be resolved.

The `is_valid_gtid()` function handles now the situation when ':' is absent, but even in this case two parts of code `Query_Processor::is_valid_gtid` and `MySQL_Session::handler___client_DSS_QUERY_SENT___server_DSS_NOT_INITIALIZED__get_connection` can be changed independently in future and cause a crash of the proxysql. The NULL return value of the `index()` functions must be handled properly.

kirs committed:
Prevent crash on index() returning NULL

in handler___client_DSS_QUERY_SENT___server_DSS_NOT_INITIALIZED__get_connection

kirs commented:
I somehow assumed that `index()` in `handler___client_DSS_QUERY_SENT___server_DSS_NOT_INITIALIZED__get_connection` was intentionally "unsafe" because at that point it's only supposed to accept valid GTIDs, but it makes sense to make it safe in both places in case it's changed in the future.

I pushed a commit that makes `index()` usage safe to NULL in that function. I hope I understand you correctly, but if not, I'm happy to follow up.

Thanks again for reviewing this.

milanatshopify commented on include/query_processor.h:
```diff
@@ -305,7 +311,7 @@ class Query_Processor {
 	void print_version();
 	void reset_all(bool lock=true);
-	void wrlock();		// explicit write lock, to be used in multi-isert 
+	void wrlock();		// explicit write lock, to be used in multi-isert
```
May as well fix the typo in missing "n" in insert :)

renecannao commented:
LGTM!
Thank you everybody!
